### PR TITLE
[random] add jax_legacy_prng_key flag

### PR DESCRIPTION
**Pull request by jakevdp**
Part of #9263

### CODE REVIEW

File: jax/_src/random.py [hunk 1]
```diff
@@ -72,7 +72,16 @@ def _check_prng_key(key) -> tuple[prng.PRNGKeyArray, bool]:
   if isinstance(key, prng.PRNGKeyArray):
     return key, False
   elif _arraylike(key):
-    if config.jax_enable_custom_prng:
+    if config.jax_raw_prng_key == 'error':
+      raise ValueError(
+        'Raw array passed as key to jax.random function. If this was intended, '
```
Comment (froystig) [inline jax/_src/random.py:77]:
```suggestion
        'uint32 array passed as key to jax.random function. If this was intended, '
```

File: jax/_src/random.py [hunk 2]
```diff
@@ -72,7 +72,16 @@ def _check_prng_key(key) -> tuple[prng.PRNGKeyArray, bool]:
...
+        'set jax_raw_prng_key="allow".')
+    elif config.jax_raw_prng_key == 'warn':
+      warnings.warn(
+        'Raw array passed as key to jax.random function. If this was intended, '
```
Comment (froystig) [inline jax/_src/random.py:81]:
```suggestion
        'uint32 array passed as key to jax.random function. If this was intended, '
```
similarly

File: jax/_src/random.py (hunk 1 above)
Comment (froystig) [reply jax/_src/random.py:77]: If we do say "uint32" here, should we *check* the dtype to ensure it's actually u32 before issuing this message? I hope we err on non-u32 inputs at some point already, so that this wouldn't be a change in error behavior...

File: jax/_src/config.py [hunk 3]
```diff
@@ -794,6 +794,13 @@ def update_thread_local_jit_state(**kw):
           'computations. Logging is performed with `logging` at WARNING '
           'level.'))
 
+raw_prng_key = config.define_enum_state(
```
Comment (froystig) [inline jax/_src/config.py:797]:
```suggestion
legacy_prng_key = config.define_enum_state(
```

I'm not sure everyone will understand what "raw" means. What do you think of this? Other terms: `uint32`, or `numeric`. See other comments, relatedly.

**Review by froystig — COMMENTED** [review]
Nice! See comments on "raw" as a term. What do we think? Only raising this since it's a public API.

File: jax/_src/random.py (hunk 1 above)
Comment (jakevdp) [reply jax/_src/random.py:77]: We error on non-u32 inputs in `prng.random_wrap()`, which is called below these statements.

What if I say "Legacy key array" rather than "Raw array" here, and add some more context about using `jax.random.key()`?

File: jax/_src/config.py (hunk 3 above)
Comment (jakevdp) [reply jax/_src/config.py]: Changed to `legacy` – thanks

File: jax/_src/random.py (hunk 1 above)
Comment (froystig) [reply jax/_src/random.py]: Since we error here under the `'error'` option, we'd mask the underlying `random_wrap` error. So let's validate here upfront, then proceed to respect the new flag?

**Review by froystig — APPROVED** [review]

File: jax/_src/random.py (hunk 1 above)
Comment (jakevdp) [reply jax/_src/random.py]: Done - also added a test to confirm that the key correctness error takes precedence.
